### DISCUSSION TITLE
Sparse MSIX package starts correctly with MSIX Hero, but not from File Explorer

### DISCUSSION THREAD

**Discussion by kainjow** [marcinotorowski/MSIX-Hero#128] [Q&A]
I'm putting together a [sparse MSIX](https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps) package and MSIX Hero has been great for basic troubleshooting and quick uninstalling. Thanks for the utility!

I thought I'd post here to see if anyone can help troubleshoot an issue I'm seeing with general MSIX development.

When the exe launches through Explorer, it doesn't report a package identity (via [GetCurrentPackageFullName](https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentpackagefullname)) which is required for WinRT APIs. Yet, when I use MSIX Hero to start it, it **does** report identity. I've been looking through everything related to app manifests (which I believe is already correct as it launches correctly through MSIX Hero) and other things, and can't find much on this. Any help would be appreciated.

**Comment by DrusTheAxe** [comment]
Is MSIX Hero launching the app by CreateProcess (or ShellExecute) with the exe filename (like Explorer)? Or by calling Khttps://docs.microsoft.com/en-us/windows/win32/api/shobjidl_core/nf-shobjidl_core-iapplicationactivationmanager-activateapplication (like when you click on a tile in StartMenu)? The former just runs an exe creating process; you won't get package identity. The latter launcches an app thus creating a process with package identity etc.

Unless...

If you add `<msix...>` to a side-by-side manifest in your exe thne CreateProcess will see the process to be created won't have package identity BUT WAIT! The exe's embedded SxS manifest provides application identity information. If your exe has this element AND it corresponds to a package registered for the user then CreateProcess recognizes it's an application and creates a process with identity. Same net result as `ActivateApplication()`.

See https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/grant-identity-to-nonpackaged-apps and https://blogs.windows.com/windowsdeveloper/2019/10/29/identity-registration-and-activation-of-non-packaged-win32-apps/ for more information.

Alternatively, you can declare an [AppExecutionAlias](https://docs.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-uap5-appexecutionalias) in your `appxmanifest.xml` e.g. `<uap5:AppExecutionAlias>foo.exe</>` and then you can run "foo.exe" from a command line and it'll launch your app. Again, different path but same net result as `ActivateAppliccation()`.

**Comment by marcinotorowski** [comment]
MSIX Hero runs packages using a specially crafted entry point and shell execute flag set to true. This way, the process receives correct identity. The key part is a special prefix `shell:appsFolder` followed by package family name and application ID to start. This is also what a standard shortcuts do, for example:

![image](https://user-images.githubusercontent.com/22944216/198867713-32baa4bd-3048-4630-a411-5b8d4d8b1a32.png)

The relevant code for this is here: https://github.com/marcinotorowski/MSIX-Hero/blob/5b389f8b657c35f86cab2a4c7526b4f92eb003c8/src/Otor.MsixHero.Appx/Packaging/Services/AppxPackageRunService.cs#L168

See also https://ss64.com/nt/shell.html
